drivers/ethernet/eth_native_posix.c

```diff
@@ -0,0 +1,268 @@
+	bool init_done;
+	bool status;
+};
+
+NET_STACK_DEFINE(RX_ZETH, eth_rx_stack, CONFIG_ETH_NATIVE_POSIX_RX_STACK_SIZE,
+		 CONFIG_ETH_NATIVE_POSIX_RX_STACK_SIZE);
+static struct k_thread rx_thread_data;
+
+/* TODO: support multiple interfaces */
+static struct eth_context eth_context_data;
+
+static struct eth_context *get_context(struct net_if *iface)
+{
+	return net_if_get_device(iface)->driver_data;
+}
+
+static int eth_send(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct eth_context *ctx = get_context(iface);
+	struct net_buf *frag;
+	int count = 0;
+
+	/* First fragment contains link layer (Ethernet)
+	 * headers so send it first.
+	 */
+	count = net_pkt_ll_reserve(pkt) + pkt->frags->len;
+	memcpy(ctx->send, net_pkt_ll(pkt), count);
+
+	/* Then the remaining data */
+	frag = pkt->frags->frags;
+	while (frag) {
+		memcpy(ctx->send + count, frag->data, frag->len);
+		count += frag->len;
+		frag = frag->frags;
+	}
+
+	net_pkt_unref(pkt);
+
+	eth_write_data(ctx->dev_fd, ctx->send, count);
+
+	return 0;
+}
+
+#if !defined(CONFIG_ETH_NATIVE_POSIX_RANDOM_MAC)
+static int mac_addr_from_str(struct eth_context *ctx, const char *src)
+{
+	unsigned int len, i;
+	char *endptr;
+
+	len = strlen(src);
+	for (i = 0; i < len; i++) {
+		if (!(src[i] >= '0' && src[i] <= '9') &&
+		    !(src[i] >= 'A' && src[i] <= 'F') &&
+		    !(src[i] >= 'a' && src[i] <= 'f') &&
+		    src[i] != ':') {
+			return -EINVAL;
+		}
+	}
+
+	memset(ctx->mac_addr, 0, sizeof(ctx->mac_addr));
+
+	for (i = 0; i < sizeof(ctx->mac_addr); i++) {
+		ctx->mac_addr[i] = strtol(src, &endptr, 16);
+		src = ++endptr;
+	}
+
+	return 0;
+}
+#endif
+
+static int eth_init(struct device *dev)
+{
+	struct eth_context *ctx = dev->driver_data;
+
+	ARG_UNUSED(ctx);
+
+	SYS_LOG_DBG("[%p] dev %p", ctx, dev);
+
+	return 0;
+}
+
+static struct net_linkaddr *eth_get_mac(struct eth_context *ctx)
+{
+	ctx->ll_addr.addr = ctx->mac_addr;
+	ctx->ll_addr.len = sizeof(ctx->mac_addr);
+
+	return &ctx->ll_addr;
+}
+
+static int read_data(struct eth_context *ctx, int fd)
+{
+	struct net_pkt *pkt;
+	struct net_buf *frag;
+	int ret;
+
+	ret = eth_read_data(fd, ctx->recv, sizeof(ctx->recv));
+	if (ret > 0) {
+		pkt = net_pkt_get_reserve_rx(0, NET_BUF_TIMEOUT);
+		if (!pkt) {
+			return -ENOMEM;
+		}
+
+		do {
+			int count = 0;
+
+			frag = net_pkt_get_frag(pkt, NET_BUF_TIMEOUT);
+			if (!frag) {
+				net_pkt_unref(pkt);
+				return -ENOMEM;
+			}
+
+			net_pkt_frag_add(pkt, frag);
+
+			net_buf_add_mem(frag, ctx->recv + count,
+					min(net_buf_tailroom(frag), ret));
+			ret -= frag->len;
+			count += frag->len;
+		} while (ret > 0);
+
+		SYS_LOG_DBG("Recv pkt %p len %d", pkt, net_pkt_get_len(pkt));
+
+		if (net_recv_data(ctx->iface, pkt) < 0) {
+			net_pkt_unref(pkt);
+		}
+	}
+
+	return 0;
+}
+
+static void eth_rx(struct eth_context *ctx)
+{
+	int ret;
+
+	NET_DBG("Starting ZETH RX thread (stack %d bytes)",
+		CONFIG_ETH_NATIVE_POSIX_RX_STACK_SIZE);
+
+	while (1) {
+		if (net_if_is_up(ctx->iface)) {
```

Comment:
Instead of polling all interfaces, why not select() on all of them at once?

Reply:
the driver does not support yet multiple iface it seems (line 57)

Reply:
Hmm, I am not sure I understood the question. Currently this eth driver supports only one TAP device. If you mean that we could have NULL timeout when calling select(), then that would not work as then other part of the system would not get chance to run.

Reply:
> the driver does not support yet multiple iface it seems (line 57)

No, the VLAN PR at #5688 would need to go in first as it has various things to support multiple ethernet interfaces. This driver can then enhanced to support multiple interfaces (I actually had TODO comment in the beginning of the file)